src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -20,94 +20,130 @@
-
-                try {
-                    key = getPublicKey(decoded, "RSA");
-                } catch (Exception e) {
-                    log.debug("No public RSA key, try other algos ({})", e.toString());
-                }
-
-                try {
-                    key = getPublicKey(decoded, "EC");
-                } catch (Exception e) {
-                    log.debug("No public ECDSA key, try other algos ({})", e.toString());
-                }
-
-                if(key != null) {
-                    jwtParser = Jwts.parser().setSigningKey(key);
-                } else {
-                    jwtParser = Jwts.parser().setSigningKey(decoded);
-                }
+            final String minimalKeyFormat = signingKey
+                    .replace("-----BEGIN PUBLIC KEY-----\n", "")
+                    .replace("-----END PUBLIC KEY-----", "");
+
+            System.out.println("THIS IS THE minimalKeyFormat of signingKey: " + minimalKeyFormat);
+
+            final byte[] decoded = Base64.getDecoder().decode(minimalKeyFormat);
+            Key key = null;
+
+            try {
+                key = getPublicKey(decoded, "RSA");
+            } catch (Exception e) {
+                log.debug("No public RSA key, try other algos ({})", e.toString());
+            }
 
+            try {
+                key = getPublicKey(decoded, "EC");
+            } catch (Exception e) {
+                log.debug("No public ECDSA key, try other algos ({})", e.toString());
             }
+
+            if (Objects.nonNull(key)) {
+                return Jwts.parser().setSigningKey(key);
+            }
+            // Fallback to the decoded signing key
+            // TODO: Should we ever do this, I think no??
+            return Jwts.parser().setSigningKey(signingKey);
         } catch (Throwable e) {
             log.error("Error while creating JWT authenticator", e);
+            e.printStackTrace();
             throw new RuntimeException(e);
         }
+    }
+
+    private List<String> extractSecurityRolesFromClaims(Claims claims) {
+        Object rolesObject = ObjectUtils.firstNonNull(claims.get("er"), claims.get("dr"));
+        List<String> roles;
+
+        if (rolesObject == null) {
+            log.warn(
+                    "Failed to get roles from JWT claims. Check if this key is correct and available in the JWT payload.");
+            roles = List.of();
+        } else {
+            final String rolesClaim = rolesObject.toString();
+
+            // Extracting roles based on the compatbility mode
+            String decryptedRoles = rolesClaim;
+            if (rolesObject == claims.get("er")) {
+                decryptedRoles = EncryptionDecryptionUtil.decrypt(encryptionKey, rolesClaim);
+            }
+            roles = Arrays.stream(decryptedRoles.split(",")).map(String::trim).collect(Collectors.toList());
+        }
+
+        return roles;
+    }
+
+    private String[] extractBackendRolesFromClaims(Claims claims) {
```

Comment:
IMO We shouldn't add backend roles in this token - discussion is in https://github.com/opensearch-project/security/issues/2865

Reply:
Perhaps we can just combine the roles at this stage and then drop the backenedRoles from the later AuthCredentials usage.

Reply:
Ya I have changed it, align with our discussion on #2865 that only including backend roles when the bwc mode is on.

Reply:
For obo tokens that are ferried to an extension when handling a REST request then backend roles should be included if the plugin being converted to an extension relies on them. For obo tokens generated during the call to the CreateOboToken endpoint I don't think they should be included either. 

Since some obo tokens may have them as a claim then this method should remain.